[재연] 문자열 덧셈 계산기

csbsjy opened the pull request:
두번째 질문입니당 
![image](https://user-images.githubusercontent.com/47847993/75608125-1cf3dd80-5b40-11ea-9683-d60b99b5143c.png)

코드를 이렇게 수정했는데요 두가지 질문있어욤
1. 사실 반환값이 Optional 이라 orElseThrow 를 하긴 했는데.. 실제로는 저기 null 이 있을 일이 없거든요 앞에서 다 거르기 때문에.. 그래도 저기서 던지는게 맞을까요? 처음에 .get().getValue() 하니까 좀 어색한 느낌이있어서여

2. reduce 사용법을 잘 몰라서 제가 이상하게 짠거같긴한데, 저렇게 짜는게 맞나여? reduce 사용하면 짧아진다구 하셨는데 어째 더 길어진거같아서염 ..!  

===============================================


코드는 CalculatorServiceTest 를 시작으로 봐주시면 좋을 것 같습니다.

한가지 고민인점은, Splitter 클래스에서
![image](https://user-images.githubusercontent.com/47847993/75412546-26295280-5966-11ea-9927-ba5e58728163.png)

위 클래스변수에 대해 어떻게 생각하시는지요?

static final 에는 기본형, 불변클래스를 사용하라고 이펙티브자바에서 본 것 같은데 패턴 클래스도 까보니까 스레드세이프한 애 같아서 했거든여.

해도 문제 없다면 위와 같은 변수는 상수가 아니니까 대문자와 _ 를 섞어쓰는 컨벤션을 따르지 않아도 되는걸까욤? 


+ final 이 붙으면 상수. Pattern 도 상수가 된 것이기 때문에 대문자 _ 컨벤션을 따라야한다.

csbsjy committed:
1st commit
1. Calculator, CalculatorService, Splitter로 책임 분리
2. TestCode 작성 완료

csbsjy committed:
[2nd Commit]
1. Split Class Matcher 로 로직 변경
2. 테스트코드 변경
3. Pattern 처럼 스레드세이프한 객체의 상수화. 네이밍은 대문자? 소문자?

csbsjy committed:
[3rd commit]
1. 서비스 불필요한 validation 제거 및 테스트 수정

csbsjy committed:
codeconvention

csbsjy committed:
Codeconvention, double > long 수정

1. Double > Long 안바뀌어있던 것 수정

jyami-kim commented on src/main/java/domain/Calculator.java:
```diff
@@ -0,0 +1,14 @@
+import java.util.List;
+
+public class Calculator {
+    public double calculate(List<String> operands) {
```
```suggestion
    public long calculate(List<String> operands) {
```

mapToLong 하고 double 이길래 한번 넣어봐여 (suggestion 테스트)

csbsjy replied:
Thankyou <3

chldbtjd2272 commented on src/main/java/support/Splitter.java:
```diff
@@ -0,0 +1,43 @@
+    private final static int CUSTOM_DELIMITER_INDEX = 1;
+    private final static int REAL_FORMULA_INDEX = 2;
+
+    private final static Pattern customDelimiterPattern = Pattern.compile("//(.*)₩n(.*)");
```
final이 붙은 것들을 보통 상수라고해
static final로 시작하는 컨밴션은 대문자를 사용하는게 일반적인 컨밴션이고,
위와같이 패턴을 정의하는건 좋아! Pattern을 컴파일하는데 cpu가 많이 소모되니까 저런식으로 만드는게 일반적인 Pattern사용법이야

chldbtjd2272 commented on src/main/java/domain/dto/Operand.java:
```diff
@@ -0,0 +1,26 @@
+
+    private void checkNegative(long parsedOperand) {
+        if (parsedOperand < 0) {
+            throw new IllegalArgumentException(String.format("음수인 %s 값은 들어올 수 없습니다", parsedOperand));
```
메세지가 제대로 안나올거같은데 문자열만 %s를 사용해

chldbtjd2272 commented on src/main/java/domain/Calculator.java:
```diff
@@ -0,0 +1,14 @@
+public class Calculator {
+    public long calculate(List<String> operands) {
+        return operands.stream()
+                .map(Operand::new)
```
reduce를 사용하면 한줄로 가능할거 같고 Operand의 값을 꺼내서 하는것보단 새로운 두개의 값을 더해서 새로운 오퍼랜드를 꺼내는게 좋을듯

csbsjy replied:
선생님 reduce 너무 어려워요

chldbtjd2272 commented on src/test/java/domain/CalculatorTest.java:
```diff
@@ -0,0 +1,51 @@
+        assertThat(result).isEqualTo(6);
+    }
+
+    @DisplayName("숫자가 아닌 값이 들어갔을 때 에러를 던져준다")
```
요 에러던지는 테스트는 Parameter를 사용해서 하나의 테스트로 만들수 있을듯, 저 문자열리스트와, 오류메세지를 인자로 주면!

chldbtjd2272 commented on src/test/java/service/CalculatorServiceTest.java:
```diff
@@ -0,0 +1,67 @@
+
+    CalculatorService calculatorService = new CalculatorService(new Calculator());
+
+    @DisplayName("올바른 수식 계산 테스트")
```
요 이름은 어떤수식을 넣으면 더한결과가 나오는지를 써주면 좋을거같아 
테스트는 이클래스의 사용법에 대한 문서가 되야하는데 수식이 복잡해 어떻게 사용하는 클래스인지 잘모르겠어

chldbtjd2272 commented on src/test/java/service/CalculatorServiceTest.java:
```diff
@@ -0,0 +1,67 @@
+    }
+
+
+    @DisplayName("올바르지 않은 수식 테스트")
```
올바르지 않은 수식은 뭐지?

csbsjy replied:
Display 네임에 적으니 너무 구구절절이라 익셉션메시지를 ㄹ추가했응

chldbtjd2272 commented on src/test/java/support/SplitterTest.java:
```diff
@@ -0,0 +1,80 @@
+
+
+
+    @DisplayName("// \n 사이에 구분자가 없을 때 에러를 던져주는 지")
```
요것도 두개테스트를 하나로 합치는게가능할듯

chldbtjd2272 commented on src/main/java/service/CalculatorService.java:
```diff
@@ -0,0 +1,22 @@
+        this.calculator = calculator;
+    }
+
+    public long calculate(final String formula) {
```
요구사항 1번 빈문자열은 0리턴, 2번 숫자 하나를 입력받으면 해당숫자반환은 없는거같은데,
위와같이 요구사항이 명확한 경우에는 테스트 코드로 해당 상황들을 먼저 정의하고 실제 구현을 하는게 좋을듯 (TDD)

csbsjy replied:
T..D..D..

csbsjy committed:
1. 요구사항 추가
2. Calculator logic reduce로 변경 -> 질문
3. testcode 수정

Operand::sum 오타수정

Pattern convention

csbsjy commented:
두번째 질문입니당 
![image](https://user-images.githubusercontent.com/47847993/75608125-1cf3dd80-5b40-11ea-9683-d60b99b5143c.png)

코드를 이렇게 수정했는데요 두가지 질문있어욤
1. 사실 반환값이 Optional 이라 orElseThrow 를 하긴 했는데.. 실제로는 저기 null 이 있을 일이 없거든요 앞에서 다 거르기 때문에.. 그래도 저기서 던지는게 맞을까요? 처음에 .get().getValue() 하니까 좀 어색한 느낌이있어서여

2. reduce 사용법을 잘 몰라서 제가 이상하게 짠거같긴한데, 저렇게 짜는게 맞나여? reduce 사용하면 짧아진다구 하셨는데 어째 더 길어진거같아서염 ..!  글고 흠 Operand 한테 sum 이라는 로직을 주는게 어색한거같기두하고염 ..

csbsjy committed:
Operand 생성자 private

csbsjy committed:
Operand Test 코드 생성

csbsjy committed:
SplitTest 수정

csbsjy committed:
정적 Util 클래스는 기본생성자를 private으로 막자

jyami-kim commented on src/main/java/service/CalculatorService.java:
```diff
@@ -0,0 +1,28 @@
+        }
+        List<String> splitedFormula = Splitter.split(formula.trim());
+
+        return calculator.calculate(splitedFormula);
```
Splitter.split(formula.trim()) 자체가 splited formula 라는 의미를 담고있는데, 변수명을 굳이 써야할까..?!!
아래 리턴에 같이 합쳐도 될것같음

csbsjy replied:
오호 그러넹 
혹시 가독성 떨어질줄알았는데 바꾸고보니까 생각보다 깔끔하넹 ㅎㅅㅎ 반영했오 땡큐 !

csbsjy committed:
Splitter.split 을 변수로 빼지 않아보자

pci2676 commented on src/main/java/domain/Calculator.java:
```diff
@@ -0,0 +1,15 @@
+        return operands.stream()
+                .map(Operand::new)
+                .reduce(Operand::sum)
+                .orElseThrow(() -> new IllegalArgumentException("sum 할 수 있는 operand가 없습니다"))
```
`reduce`할때 첫번째 인자로 0을 주면 `Optional`한 결과가 나오지 않아!
```java
Integer sum = integers.reduce(0, Integer::sum); 
```

csbsjy replied:
선생님 제 코드는 지금 Operand 를 반환하눈데용, 그럼 혹시 new Operand() 를 해서 0 을 가지고 있는 애를 반환하는게 좋을까? 지금은 디폴드생성자가 없는 상태이긴행 1

csbsjy replied:
하긴 어차피 Null 일 일이 없으니까 한줄로 끝내는게 나아보이긴한다!

pci2676 commented on src/main/java/domain/dto/Operand.java:
```diff
@@ -0,0 +1,49 @@
+package domain.dto;
+
+public class Operand {
```
ㅎㅎ 이왕 `VO`로 만든거 `final`로 만들어 주는건 어떨까!
상속받을일도 없을거 같기도 하고..!!
근데 나도 안했던거같아..

csbsjy replied:
ㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋ

pci2676 commented on src/main/java/domain/dto/Operand.java:
```diff
@@ -0,0 +1,49 @@
+        return new Operand(this.value + operand.value);
+    }
+
+    private void checkNegative(long parsedOperand) {
```
`private` 메서드는 그와 관련있는 `public` 메서드 근처에 위치해서 짝을 이루는게 읽기 좋다고 생각해!
생성자에서 쓰이는 메서드니까 그 아래에 있으면 난 좋을거 같아

pci2676 commented on src/main/java/support/Splitter.java:
```diff
@@ -0,0 +1,46 @@
+    }
+
+    private static void validateDelimiter(final String customDelimiter) {
+        if (customDelimiter.matches(NUMERIC_REGEX) || customDelimiter.isEmpty()) {
+            throw new IllegalArgumentException(String.format("%s 는(은) 올바르지 않은 커스텀구분자입니다.", customDelimiter));
+        }
```
헉 구분자가 숫자가 되는 경우에 대해서도 유효성을 잡아줬구나
난 생각치도 못했어..!

csbsjy committed:
1. Operand 기본생성자 추가, reduce 수정
2. final static -> static final

csbsjy committed:
@NullAndEmptySource 를 알았다

malibinYun commented on src/main/java/service/CalculatorService.java:
```diff
@@ -0,0 +1,27 @@
+            return 0;
+        }
+
+        return calculator.calculate(Splitter.split(formula.trim()));
```
이 구문은 한줄에 .이 너무 많아서 가독성이 안좋아보여요.
Splitter.split(formula.trim()) 이 친구를 한 번 래핑하는 편이 더 나아보입니다.

csbsjy replied:
오홍 의견이갈리네여 ~~  아래 의견도 있었었습니당 ㅋㅋㅋ
다시 고민해볼게여! 
![image](https://user-images.githubusercontent.com/47847993/76053999-76319600-5fb2-11ea-9a94-814663377c33.png)
